server/src/main/java/org/elasticsearch/index/mapper/NestedPathFieldMapper.java

```diff
@@ -0,0 +1,181 @@
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.fielddata.plain.ConstantIndexFieldData;
+import org.elasticsearch.index.query.QueryShardContext;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+public class NestedPathFieldMapper extends MetadataFieldMapper {
+
+    public static final String NAME_PRE_V8 = "_type";
+    public static final String NAME = "_nested_path";
+
+    public static String name(Settings settings) {
+        if (Version.indexCreated(settings).before(Version.V_8_0_0)) {
+            return NAME_PRE_V8;
+        }
+        return NAME;
+    }
+
+    public static Query filter(Settings settings, String path) {
+        return new TermQuery(new Term(name(settings), new BytesRef(path)));
+    }
+
+    public static Field field(Settings settings, String path) {
+        return new Field(name(settings), path, Defaults.FIELD_TYPE);
+    }
+
+    public static class Defaults {
+
+        public static final MappedFieldType FIELD_TYPE = new NestedPathFieldType();
+
+        static {
+            FIELD_TYPE.setIndexOptions(IndexOptions.DOCS);
+            FIELD_TYPE.setTokenized(false);
+            FIELD_TYPE.setStored(false);
+            FIELD_TYPE.setOmitNorms(true);
+            FIELD_TYPE.setIndexAnalyzer(Lucene.KEYWORD_ANALYZER);
+            FIELD_TYPE.setSearchAnalyzer(Lucene.KEYWORD_ANALYZER);
+            FIELD_TYPE.freeze();
+        }
+    }
+
+    public static class TypeParser implements MetadataFieldMapper.TypeParser {
+        @Override
+        public MetadataFieldMapper.Builder<?,?> parse(String name, Map<String, Object> node,
+                                                      ParserContext parserContext) throws MapperParsingException {
+            throw new MapperParsingException(name(parserContext.mapperService().getIndexSettings().getSettings()) + " is not configurable");
+        }
+
+        @Override
+        public MetadataFieldMapper getDefault(MappedFieldType fieldType, ParserContext context) {
+            final IndexSettings indexSettings = context.mapperService().getIndexSettings();
+            return new NestedPathFieldMapper(indexSettings, fieldType);
+        }
+    }
+
+    public static final class NestedPathFieldType extends StringFieldType {
+
+        NestedPathFieldType() {
+        }
+
+        protected NestedPathFieldType(NestedPathFieldType ref) {
+            super(ref);
+        }
+
+        @Override
+        public MappedFieldType clone() {
+            return new NestedPathFieldType(this);
+        }
+
+        @Override
+        public String typeName() {
+            return NAME;
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName) {
+            Function<MapperService, String> typeFunction = mapperService -> mapperService.documentMapper().type();
```

Comment:
should we disallow fielddata on this field ? It's an implementation details and I am not sure that we want to expose it in aggs or sorting ?

Reply:
Maybe keep it only for indices created before 8 to handle bwc ?

Reply:
I'll remove it - indices created before 8 still use TypeFieldMapper, so BWC is handled that way

Reply:
I pushed 7ea0da8